src/utils/EventUtils.ts

```diff
@@ -0,0 +1,227 @@
+  const keys = Object.keys(args).filter((key: string) => isNaN(+key)); // Extract non-numeric keys.
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const returnedObject: any = {};
+  keys.forEach((key: string) => {
+    switch (typeof args[key]) {
+      case "boolean": // fallthrough
+      case "number":
+      case "string":
+        returnedObject[key] = args[key];
+        break;
+      case "object":
+        if (Array.isArray(args[key])) {
+          returnedObject[key] =
+            Object.keys(args[key].filter((key: string) => isNaN(+key))).length > 0
+              ? spreadEvent(args[key]) // Record/array hybrid...
+              : args[key]; // Just an array
+        } else {
+          returnedObject[key] = args[key];
+        }
+        break;
+    }
+  });
+
+  // ID information, if included in an event, should be cast to a number rather than a BigNumber.
+  if (returnedObject.groupIndex) {
+    returnedObject.groupIndex = Number(returnedObject.groupIndex);
+  }
+  if (returnedObject.leafId) {
+    returnedObject.leafId = Number(returnedObject.leafId);
+  }
+  if (returnedObject.chainId) {
+    returnedObject.chainId = Number(returnedObject.chainId);
+  }
+  if (returnedObject.destinationChainId) {
+    returnedObject.destinationChainId = Number(returnedObject.destinationChainId);
+  }
+  if (returnedObject.originChainId) {
+    returnedObject.originChainId = Number(returnedObject.originChainId);
+  }
+  if (returnedObject.repaymentChainId) {
+    returnedObject.repaymentChainId = Number(returnedObject.repaymentChainId);
+  }
+  if (returnedObject.l2ChainId) {
+    returnedObject.l2ChainId = Number(returnedObject.l2ChainId);
+  }
+  if (returnedObject.rootBundleId) {
+    returnedObject.rootBundleId = Number(returnedObject.rootBundleId);
+  }
+
+  return returnedObject;
+}
+
+export interface EventSearchConfig {
+  fromBlock: number;
+  toBlock: number;
+  maxBlockLookBack?: number;
+}
+
+export async function paginatedEventQuery(
+  contract: Contract,
+  filter: EventFilter,
+  searchConfig: EventSearchConfig,
+  retryCount = 0
+): Promise<Event[]> {
+  // If the max block look back is set to 0 then we dont need to do any pagination and can query over the whole range.
+  if (searchConfig.maxBlockLookBack === 0) {
+    return await contract.queryFilter(filter, searchConfig.fromBlock, searchConfig.toBlock);
+  }
+
+  // Compute the number of queries needed. If there is no maxBlockLookBack set then we can execute the whole query in
+  // one go. Else, the number of queries is the range over which we are searching, divided by the maxBlockLookBack,
+  // rounded up. This gives us the number of queries we need to execute to traverse the whole block range.
+  const paginatedRanges = getPaginatedBlockRanges(searchConfig);
+
+  try {
+    return (
+      (
+        await Promise.all(
+          paginatedRanges.map(([fromBlock, toBlock]) => contract.queryFilter(filter, fromBlock, toBlock))
+        )
+      )
```

Comment:
Changed Promise.map from bluebird into `promise.all(...)`

Reply:
We should circle back and see whether we still need Bluebird in relayer-v2 after this migration.